src/config.js

```diff
@@ -23,6 +23,33 @@
  */
 const env = self.AMP_CONFIG || {};
 
+/**
+ * Check for a custom URL definition in special <meta> tags. Note that this does
+ * not allow for distinct custom URLs in AmpDocShadow instances. The shell is
+ * allowed to define one set of custom URLs via AMP_CONFIG (recommended) or by
+ * including <meta> tags in the shell <head>. Those custom URLs then apply to
+ * all AMP documents loaded in the shell.
+ * @param {string} name
+ * @return {?string}
+ * @private
+ */
+function getMetaUrl(name) {
+  // Avoid exceptions in unit tests
+  if (!self.document || !self.document.head) {
+    return null;
+  }
+
+  const metaEl = self.document.head./*OK*/ querySelector(
```

Comment:
This will immediately invoke the meta lookup, which creates a race between head parsing and JS loading. We need to delay this as much as possible.

Reply:
Thanks for the review. I see your concern. I'll hunt around for a suitable event on which to listen or a flag that has been set. Notably, though, I'm trying to avoid adding dependencies to `config.js` as they quickly lead to circular dependencies when this file is imported elsewhere. I'll post back when I have a solution.

Reply:
@jridgewell I'm not sure how to accomplish the delay without turning `urls.cdn` into a promise or letting `urls.cdn` contain an incorrect value for a short time until `<head>` is fully available, and then re-evaluating its value. I'm not thrilled about either solution. The former requires refactoring all consumers of `urls.cdn` and the later is... sloppy. Could we simply make it a stipulation that `<meta name="self-host">` must appear earlier than all `<script>` tags? It's not much different than what is required for PWAs that define `AMP_CONFIG={cdnUrl:"..."}` before any AMP scripts load.

Reply:
> Could we simply make it a stipulation that `<meta name="runtime-host">` must appear earlier than alll `<script>` tags?

That seems fine to me.

@ampproject/wg-caching: Is it possible to specify that this meta tag must come before the `<script src=v0.js>`?

Reply:
AFAIK, validator doesn't currently support ordering constraints, though it seems possible to change that. I'll defer to @Gregable and @honeybadgerdontcare on whether we should.

A half-way solution is to change the [reorder head transformer](https://github.com/ampproject/amppackager/blob/releases/transformer/transformers/reorderhead.go) to ensure this is correct on AMP cache, at least. We could also update [AMP optimizer](https://github.com/ampproject/amp-toolbox/blob/master/packages/optimizer/lib/transformers/ReorderHeadTransformer.js).

Reply:
We actually want Caches to strip the meta entirely. Adding this to Optimizer would be good, though.

Reply:
Ordering of meta tags by optimizer and head transformer is being implemented/discussed here: https://github.com/ampproject/amp-toolbox/pull/636 . It makes sense to split out the reordering into a separate PR from the geoapi change, so I'll do that next.

Reply:
For the Google AMP Cache and the AMP Packager, we do put `meta` tags before all `script` tags in the [reorder head transformer](https://github.com/ampproject/amppackager/blob/releases/transformer/transformers/reorderhead.go). This is a requirement for several extensions. We also only keep an explicit subset of meta tags, see [internal](http://go/amphtml-transformers/t/MetaTag) or [external](https://github.com/ampproject/amphtml/blob/master/spec/amp-cache-modifications.md#remove-non-whitelisted-meta-tags) documentation.

It looks like the AMP optimizer does deviate from what we do, with putting the runtime before other meta tags. That may be okay but guidance I've received is all `meta` tags should go before any `script` tag. @sebastianbenz 

As to restricting the ordering @twifkak is right that we don't have a general way to do that but we have some rudimentary way to do it, see [ChildTagSpec](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L347). That won't work here. We can file a feature request but it would need to be implemented first to avoid making pages already implementing this `meta` tag as invalid.

Reply:
> We actually want Caches to strip the meta entirely.

@jridgewell what's the reason for this? The cache doesn't strip all `meta` tags either.  Not sure it's a good idea for Optimizer to remove meta tags as it might break publishers on the origin.

I've filed https://github.com/ampproject/amp-toolbox/issues/641 for updating optimizer to correctly place meta tags.

> A half-way solution is to change the reorder head transformer to ensure this is correct on AMP cache, at least. We could also update AMP optimizer.

This sounds like a valid approach as long as runtime self-hosting is only supported by `transformed` AMP.

Reply:
Optimizer shouldn't, only cache. But actually, the code now only respects the meta if we're on Origin, so it won't be bad either way.

Reply:
> as long as runtime self-hosting is only supported by transformed AMP.

I would prefer this is not made a stipulation.

It's worth pointing out, the code change made by this PR ignores the meta tags under discussion if the origin is a cache origin.